Nodejs/Product/TestAdapter/TestExecutorWorker.cs

```diff
@@ -0,0 +1,390 @@
+        private List<TestCase> currentTests;
+        private ProcessOutput nodeProcess;
+        private TestResult currentResult;
+        private ResultObject currentResultObject;
+
+        public TestExecutorWorker(IRunContext runContext, IFrameworkHandle frameworkHandle)
+        {
+            this.frameworkHandle = frameworkHandle;
+            this.runContext = runContext;
+        }
+
+        public void Cancel()
+        {
+            //let us just kill the node process there, rather do it late, because VS engine process 
+            //could exit right after this call and our node process will be left running.
+            KillNodeProcess();
+            this.cancelRequested.Set();
+        }
+
+        /// <summary>
+        /// This is the equivalent of "RunAll" functionality
+        /// </summary>
+        /// <param name="sources">Refers to the list of test sources passed to the test adapter from the client.  (Client could be VS or command line)</param>
+        /// <param name="runContext">Defines the settings related to the current run</param>
+        /// <param name="frameworkHandle">Handle to framework.  Used for recording results</param>
+        public void RunTests(IEnumerable<string> sources, ITestDiscoverer discoverer)
+        {
+            ValidateArg.NotNull(sources, nameof(sources));
+            ValidateArg.NotNull(discoverer, nameof(discoverer));
+
+            this.cancelRequested.Reset();
+
+            var receiver = new TestReceiver();
+            discoverer.DiscoverTests(sources, /*discoveryContext*/null, this.frameworkHandle, receiver);
+
+            if (this.cancelRequested.WaitOne(0))
+            {
+                return;
+            }
+
+            this.RunTests(receiver.Tests);
+        }
+
+        /// <summary>
+        /// This is the equivalent of "Run Selected Tests" functionality.
+        /// </summary>
+        /// <param name="tests">The list of TestCases selected to run</param>
+        /// <param name="runContext">Defines the settings related to the current run</param>
+        /// <param name="frameworkHandle">Handle to framework.  Used for recording results</param>
+        public void RunTests(IEnumerable<TestCase> tests)
+        {
+            ValidateArg.NotNull(tests, nameof(tests));
+            this.cancelRequested.Reset();
+
+            // .ts file path -> project settings
+            var fileToTests = new Dictionary<string, List<TestCase>>();
+
+            // put tests into dictionary where key is their source file
+            foreach (var test in tests)
+            {
+                if (!fileToTests.ContainsKey(test.CodeFilePath))
+                {
+                    fileToTests[test.CodeFilePath] = new List<TestCase>();
+                }
+                fileToTests[test.CodeFilePath].Add(test);
+            }
+
+            // where key is the file and value is a list of tests
+            foreach (var entry in fileToTests)
```

Comment:
If the key isn't used, consider iterating over `fileToTests.Values`.

Reply:
done